pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -239,8 +262,32 @@ func (d *ServiceEntryStore) maybeRefreshIndexes() {
 	di := map[host.Name]map[string][]*model.ServiceInstance{}
 	dip := map[string][]*model.ServiceInstance{}
 
+	wles, err := d.store.List(workloadEntryKind, model.NamespaceAll)
+	if err != nil {
+		log.Errorf("Error listing workload entries: %v", err)
+	}
 	for _, cfg := range d.store.ServiceEntries() {
-		updateInstances(convertInstances(cfg, nil), di, dip)
+		services := convertServices(cfg)
+		updateInstances(convertInstances(cfg, services), di, dip)
+
+		se := cfg.Spec.(*networking.ServiceEntry)
+		// If we have a workload selector, we will add all instances from WorkloadEntries. Otherwise, we continue
+		if se.WorkloadSelector == nil {
+			continue
+		}
+		for _, wcfg := range wles {
+			wle := wcfg.Spec.(*networking.WorkloadEntry)
+			// We will only select entries in the same namespace
+			if wcfg.Namespace != cfg.Namespace {
```

Comment:
Just cause I'm curious, is any of this code a scalability concern?

We could do a single pass over the workload entries to put them in a per-namespace map rather than iterating over all of them for every service entry. Especially since there is now one WorkloadEntry per endpoint this could get potentially fairly big.

Also cause I'm curious, why do we do a full refresh when there are updates? Shouldn't we be watching and only updating things that changed? In a large mesh I can imagine this becoming a huge waste of CPU.

Reply:
Yeah that probably makes sense to have the ns map.

As far as the full refresh - that was a quick hack to get something working. Before this merges I'll add proper event handling, I just wanted to get some early eyes on this

Reply:
Ah OK I see, we only refresh indexes as needed, so will need more code in the event handler to do smaller updates.

Again since I'm curious, what are the conditions where we do a full refresh of the indexes? Ideally it would only happen on startup or if we somehow miss some updates right?

Reply:
If configs change (not workload entry but other configs), we do full refresh. If services are added/removed, we refresh (as listeners change). Endpoint changes are incremental. IIRC we also have the ability to do full refresh only on the impacted namespaces

Reply:
I added this optimization.